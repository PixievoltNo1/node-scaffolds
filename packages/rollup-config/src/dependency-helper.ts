import { locateNearestFilepath } from '@guanghechen/locate-helper'
import fs from 'fs-extra'

export const dependencyKeys: () => ReadonlyArray<
  'dependencies' | 'optionalDependencies' | 'peerDependencies'
> = () => ['dependencies', 'optionalDependencies', 'peerDependencies']

/**
 * Collect all dependencies declared in the package.json and the dependency's
 * dependencies and so on.
 *
 * @param packageJsonPath
 * @param dependenciesFields
 * @param additionalDependencies
 * @param isAbsentAllowed
 */
export function collectAllDependencies(
  packageJsonPath: string | null,
  dependenciesFields: ReadonlyArray<string>,
  additionalDependencies: ReadonlyArray<string> | null = null,
  isAbsentAllowed: ((moduleName: string) => boolean) | null = null,
): string[] {
  const result: string[] = []
  if (isAbsentAllowed == null) {
    const regex = /^@types\//
    // eslint-disable-next-line no-param-reassign
    isAbsentAllowed = moduleName => regex.test(moduleName)
  }

  /**
   * @param {string} dependency
   * @returns {void}
   */
  const followDependency = (dependency: string): void => {
    if (result.includes(dependency)) return
    result.push(dependency)

    // recursively collect
    let nextPackageJsonPath = null
    try {
      const dependencyPath = require.resolve(dependency)
      nextPackageJsonPath = locateNearestFilepath(
        dependencyPath,
        'package.json',
      )
    } catch (e) {
      switch (e.code) {
        case 'MODULE_NOT_FOUND':
          if (isAbsentAllowed!(dependency)) return
          break
        case 'ERR_PACKAGE_PATH_NOT_EXPORTED':
          return
        default:
          console.error(e)
          return
      }
    }

    if (nextPackageJsonPath == null) {
      console.warn(`cannot find package.json for '${dependency}'`)
      return
    }

    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    collectDependencies(nextPackageJsonPath)
  }

  /**
   * @param {string} dependencyPackageJsonPath
   * @returns {void}
   */
  const collectDependencies = (dependencyPackageJsonPath: string): void => {
    if (!fs.existsSync(dependencyPackageJsonPath)) {
      console.warn(`no such file or directory: ${dependencyPackageJsonPath}`)
      return
    }

    const manifest = fs.readJSONSync(dependencyPackageJsonPath)
    for (const fieldName of dependenciesFields) {
      const field = manifest[fieldName]
      if (field != null) {
        for (const dependency of Object.keys(field)) {
          followDependency(dependency)
        }
      }
    }
  }

  // collect from package.json
  if (packageJsonPath != null) {
    collectDependencies(packageJsonPath)
  }

  // collect from dependencies
  if (additionalDependencies != null) {
    for (const dependency of additionalDependencies) {
      followDependency(dependency)
    }
  }

  return result
}

export default collectAllDependencies
