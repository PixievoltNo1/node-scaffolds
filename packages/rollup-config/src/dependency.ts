import fs from 'fs-extra'
import path from 'path'

export type IDependencyField = 'dependencies' | 'optionalDependencies' | 'peerDependencies'

/**
 * Default Dependency fields
 */
export const getDefaultDependencyFields = (): IDependencyField[] => [
  'dependencies',
  'optionalDependencies',
  'peerDependencies',
]

/**
 * Collect all dependencies declared in the package.json and the dependency's dependencies and so on.
 *
 * @param packageJsonPath
 * @param dependenciesFields (such as ['dependencies', 'devDependencies'])
 * @param additionalDependencies
 * @param isAbsentAllowed
 * @returns
 */
export function collectAllDependencies(
  packageJsonPath: string | null,
  dependenciesFields: ReadonlyArray<IDependencyField> = getDefaultDependencyFields(),
  additionalDependencies: ReadonlyArray<string> | null = null,
  isAbsentAllowed: ((moduleName: string) => boolean) | null = null,
): string[] {
  const result: string[] = []
  if (isAbsentAllowed == null) {
    const regex = /^@types\//
    // eslint-disable-next-line no-param-reassign
    isAbsentAllowed = moduleName => regex.test(moduleName)
  }

  const followDependency = (dependency: string): void => {
    if (result.includes(dependency)) return
    result.push(dependency)

    // recursively collect
    let nextPackageJsonPath = null
    try {
      const dependencyPath = require.resolve(dependency)
      nextPackageJsonPath = locateNearestFilepath(dependencyPath, 'package.json')
    } catch (e: any) {
      switch (e.code) {
        case 'MODULE_NOT_FOUND':
          if (isAbsentAllowed!(dependency)) return
          break
        case 'ERR_PACKAGE_PATH_NOT_EXPORTED':
          return
        default:
          console.error(e)
          return
      }
    }

    if (nextPackageJsonPath == null) {
      console.warn(`cannot find package.json for '${dependency}'`)
      return
    }

    collectDependencies(nextPackageJsonPath)
  }

  /**
   * @param {string} dependencyPackageJsonPath
   * @returns {void}
   */
  const collectDependencies = (dependencyPackageJsonPath: string): void => {
    if (!fs.existsSync(dependencyPackageJsonPath)) {
      console.warn(`no such file or directory: ${dependencyPackageJsonPath}`)
      return
    }

    const manifest = fs.readJSONSync(dependencyPackageJsonPath)
    for (const fieldName of dependenciesFields) {
      const field = manifest[fieldName]
      if (field != null) {
        for (const dependency of Object.keys(field)) {
          followDependency(dependency)
        }
      }
    }
  }

  // collect from package.json
  if (packageJsonPath != null) {
    collectDependencies(packageJsonPath)
  }

  // collect from dependencies
  if (additionalDependencies != null) {
    for (const dependency of additionalDependencies) {
      followDependency(dependency)
    }
  }

  return result
}

function locateNearestFilepath(currentDir: string, filenames: string | string[]): string | null {
  // Uniform parameter format.
  if (!Array.isArray(filenames)) {
    // eslint-disable-next-line no-param-reassign
    filenames = [filenames]
  }

  for (const filename of filenames) {
    const filepath = path.join(currentDir, filename)
    if (fs.existsSync(filepath)) return filepath
  }

  const parentDir = path.dirname(currentDir)
  if (parentDir === currentDir) return null

  // Recursively locate.
  return locateNearestFilepath(parentDir, filenames)
}
