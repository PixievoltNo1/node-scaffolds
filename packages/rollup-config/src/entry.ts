import type { IEntryItem, IManifestEntryFields, IRawEntryItem } from './types'

export function resolveEntryItems(manifest: IManifestEntryFields): IEntryItem[] {
  const defaultEntry: IEntryItem = {
    import: manifest.module,
    require: manifest.main,
    source: [manifest.source || []].flat(),
    types: manifest.types,
  }

  const exportsEntries: IEntryItem[] = []
  const addEntryItem = (item: IRawEntryItem): void => {
    if (item.import || item.require) {
      exportsEntries.push({
        import: item.import,
        require: item.require,
        source: [item.source || []].flat(),
        types: item.types,
      })
    }
  }

  if (manifest.exports) {
    if (typeof manifest.exports === 'string') {
      exportsEntries.push({
        ...defaultEntry,
        import: manifest.exports,
      })
    } else if (typeof manifest.exports === 'object') {
      if (manifest.exports.import || manifest.exports.require) {
        addEntryItem({
          ...manifest.exports,
          source: (manifest.exports.source as string | string[] | undefined) || manifest.source,
          types: (manifest.types as string | undefined) || manifest.types,
        })
      } else {
        for (const key of Object.keys(manifest.exports)) {
          const value = manifest.exports[key as keyof typeof manifest.exports] as IRawEntryItem
          if (key === '.') {
            addEntryItem({
              ...value,
              source: (value.source as string | string[] | undefined) || manifest.source,
              types: (value.types as string | undefined) || manifest.types,
            })
          } else {
            addEntryItem(value)
          }
        }
      }
    }
  }

  const candidateEntries: IEntryItem[] = [defaultEntry, ...exportsEntries].filter(
    item => (item.import || item.require) && item.source.length > 0,
  )

  const results: IEntryItem[] = []
  const outSet: Set<string> = new Set()
  for (const item of candidateEntries) {
    let duplicated = false
    if (item.import) {
      duplicated ||= outSet.has(item.import)
      outSet.add(item.import)
    }
    if (item.require) {
      duplicated ||= outSet.has(item.require)
      outSet.add(item.require)
    }
    if (item.types) {
      duplicated ||= outSet.has(item.types)
      outSet.add(item.types)
    }
    if (!duplicated) results.push(item)
  }

  outSet.clear()
  return results
}
