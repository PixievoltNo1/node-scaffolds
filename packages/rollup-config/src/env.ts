import type { IEnv } from './types'

export interface IRawRollupConfigEnv {
  /**
   * Whether if generate sourcemaps.
   * @default true
   */
  shouldSourceMap?: boolean
  /**
   * Whether if make all dependencies external.
   * @default true
   */
  shouldExternalAll?: boolean
}

export function resolveRollupConfigEnv(rawEnv: IRawRollupConfigEnv): IEnv {
  const forceSourceMap: boolean | undefined = convertToBoolean(process.env.ROLLUP_SHOULD_SOURCEMAP)
  const forceExternalAll: boolean | undefined = convertToBoolean(
    process.env.ROLLUP_EXTERNAL_ALL_DEPENDENCIES,
  )
  const env: IEnv = {
    sourcemap: forceSourceMap ?? rawEnv.shouldSourceMap ?? true,
    externalAll: forceExternalAll ?? rawEnv.shouldExternalAll ?? true,
  }
  return env
}

function convertToBoolean(v?: unknown): boolean | undefined {
  if (typeof v === 'boolean') return v
  if (typeof v === 'string') {
    const t = v.toLowerCase()
    if (t === 'false') return false
    if (t === 'true') return true
  }
  return undefined
}
