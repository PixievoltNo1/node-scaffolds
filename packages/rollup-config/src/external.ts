import { collectAllDependencies, getDefaultDependencyFields } from '@guanghechen/helper-npm'
import type { IEnv, IManifestWithDependencies } from './types'

const builtinModules = [
  'assert',
  'async_hooks',
  'buffer',
  'child_process',
  'cluster',
  'console',
  'constants',
  'crypto',
  'dgram',
  'dns',
  'domain',
  'events',
  'fs',
  'http',
  'http2',
  'https',
  'inspector',
  'module',
  'net',
  'os',
  'path',
  'perf_hooks',
  'process',
  'punycode',
  'querystring',
  'readline',
  'repl',
  'stream',
  'string_decoder',
  'timers',
  'tls',
  'trace_events',
  'tty',
  'url',
  'util',
  'v8',
  'vm',
  'wasi',
  'worker_threads',
  'zlib',
]

const builtinExternals: ReadonlyArray<string> = builtinModules.concat(['glob', 'sync'])
export const builtinExternalSet: ReadonlySet<string> = new Set<string>(builtinExternals)

export async function resolveExternal(
  manifest: IManifestWithDependencies,
  env: IEnv,
  additionalExternal: ((id: string) => boolean) | undefined,
): Promise<(id: string) => boolean> {
  const dependencyFields = getDefaultDependencyFields()
  let dependencies: string[] = dependencyFields.reduce((acc, key) => {
    const deps = manifest[key] as Record<string, string> | string[]
    const result: string[] = Array.isArray(deps) ? deps : Object.keys(deps || {})
    return acc.concat(result)
  }, [] as string[])

  if (env.externalAll) {
    dependencies = await collectAllDependencies(null, dependencyFields, dependencies, () => true)
  }
  const externalSet = new Set(dependencies)

  const external = (id: string): boolean => {
    if (builtinExternalSet.has(id)) return true
    if (/^node:[\w\S]+$/.test(id)) return true
    if (additionalExternal && additionalExternal(id)) return true

    const m = /^([.][\s\S]*|@[^/\s]+[/][^/\s]+|[^/\s]+)/.exec(id)
    if (m == null) return false
    return externalSet.has(m[1])
  }
  return external
}
