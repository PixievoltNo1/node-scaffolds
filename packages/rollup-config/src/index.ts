export { createRollupConfig as default } from './config'
export * from './types'
export * from './config'
export * from './option-helper'
