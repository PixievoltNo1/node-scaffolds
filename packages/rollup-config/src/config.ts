import commonjs from '@rollup/plugin-commonjs'
import json from '@rollup/plugin-json'
import nodeResolve from '@rollup/plugin-node-resolve'
import typescript from '@rollup/plugin-typescript'
import type { OutputOptions, RollupOptions } from 'rollup'
import dts from 'rollup-plugin-dts'
import builtinModules from './builtin-modules.json'
import { collectAllDependencies, getDefaultDependencyFields } from './dependency'
import { convertToBoolean, coverBoolean, isArray } from './option-helper'
import type { IRawRollupConfigEnvs, IRollupConfigEnvs, IRollupConfigOptions } from './types'

const builtinExternals: string[] = builtinModules.concat(['glob', 'sync'])

/**
 * Resolve RollupConfigEnvs
 *
 * @param rawEnv
 * @returns
 */
export function resolveRollupConfigEnvs(rawEnv: IRawRollupConfigEnvs): IRollupConfigEnvs {
  const defaultShouldSourcemap = coverBoolean(
    true,
    convertToBoolean(process.env.ROLLUP_SHOULD_SOURCEMAP),
  )
  const defaultShouldExternalAll = coverBoolean(
    true,
    convertToBoolean(process.env.ROLLUP_EXTERNAL_ALL_DEPENDENCIES),
  )

  const { shouldSourceMap = defaultShouldSourcemap, shouldExternalAll = defaultShouldExternalAll } =
    rawEnv
  return { shouldSourceMap, shouldExternalAll }
}

/**
 * Create a rollup options.
 * @param options
 */
export function createRollupConfig(options: IRollupConfigOptions): RollupOptions[] {
  const env = resolveRollupConfigEnvs(options)

  const { manifest, pluginOptions = {}, additionalPlugins = [] } = options

  const { commonjsOptions, jsonOptions, nodeResolveOptions, typescriptOptions, dtsOptions } =
    pluginOptions

  const dependencyFields = getDefaultDependencyFields()
  let dependencies: string[] = dependencyFields.reduce((acc, key) => {
    const deps = manifest[key] as Record<string, string> | string[]
    const result: string[] = isArray(deps) ? deps : Object.keys(deps || {})
    return acc.concat(result)
  }, [] as string[])
  if (env.shouldExternalAll) {
    dependencies = collectAllDependencies(null, dependencyFields, dependencies, () => true)
  }
  const externalSet = new Set(builtinExternals.concat(dependencies))

  const external = (id: string): boolean => {
    const m = /^([.][\s\S]*|@[^/\s]+[/][^/\s]+|[^/\s]+)/.exec(id)
    if (m == null) return false
    return externalSet.has(m[1])
  }

  const config: RollupOptions[] = [
    {
      input: manifest.source,
      output: [
        manifest.main && {
          file: manifest.main,
          format: 'cjs',
          exports: 'named',
          sourcemap: env.shouldSourceMap,
        },
        manifest.module && {
          file: manifest.module,
          format: 'es',
          exports: 'named',
          sourcemap: env.shouldSourceMap,
        },
      ].filter((x): x is OutputOptions | any => Boolean(x)),
      external,
      plugins: [
        nodeResolve({
          browser: true,
          preferBuiltins: false,
          ...nodeResolveOptions,
        }),
        json({
          indent: '  ',
          namedExports: true,
          ...jsonOptions,
        }),
        typescript({
          include: ['**/*.tsx', '**/*.ts'],
          ...typescriptOptions,
          compilerOptions: {
            // compilerOptions
            outDir: 'lib',
            removeComments: true,
            sourceMap: env.shouldSourceMap,

            ...typescriptOptions?.compilerOptions,

            declaration: false,
            declarationMap: false,
            declarationDir: undefined,
          },
        }),
        commonjs({
          extensions: ['.js', '.jsx', '.ts', '.tsx'],
          ...commonjsOptions,
        }),
        ...additionalPlugins,
      ],
    },
    manifest.types && [
      {
        input: manifest.source,
        output: [
          {
            file: manifest.types,
            format: 'es',
          },
        ],
        external: (id: string) => {
          if (id === manifest.name + '/package.json') return true
          return external(id)
        },
        plugins: [
          dts({
            ...dtsOptions,
            compilerOptions: {
              removeComments: false,
              sourceMap: env.shouldSourceMap,
              declaration: true,
              declarationMap: false,
              emitDeclarationOnly: true,
              ...dtsOptions?.compilerOptions,
            },
          }),
        ],
      },
    ],
  ]
    .filter((x: unknown): x is RollupOptions | RollupOptions[] => !!x)
    .flat() as RollupOptions[]
  return config
}
