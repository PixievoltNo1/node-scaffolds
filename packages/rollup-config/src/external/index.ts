import type { IEnv, IManifestWithDependencies } from '../types'
import builtinModules from './builtin-modules.json' assert { type: 'json' }
import { collectAllDependencies, getDefaultDependencyFields } from './dependency'

const builtinExternals: string[] = builtinModules.concat(['glob', 'sync'])

export async function resolveExternal(
  manifest: IManifestWithDependencies,
  env: IEnv,
): Promise<(id: string) => boolean> {
  const dependencyFields = getDefaultDependencyFields()
  let dependencies: string[] = dependencyFields.reduce((acc, key) => {
    const deps = manifest[key] as Record<string, string> | string[]
    const result: string[] = Array.isArray(deps) ? deps : Object.keys(deps || {})
    return acc.concat(result)
  }, [] as string[])

  if (env.externalAll) {
    dependencies = await collectAllDependencies(null, dependencyFields, dependencies, () => true)
  }
  const externalSet = new Set(builtinExternals.concat(dependencies))

  const external = (id: string): boolean => {
    if (/^node:[\w\S]+$/.test(id)) return true

    const m = /^([.][\s\S]*|@[^/\s]+[/][^/\s]+|[^/\s]+)/.exec(id)
    if (m == null) return false
    return externalSet.has(m[1])
  }
  return external
}
