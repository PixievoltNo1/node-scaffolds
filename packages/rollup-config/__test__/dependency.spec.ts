import { jest } from '@jest/globals'
import { desensitize, locateFixtures } from 'jest.helper'
import path from 'node:path'
import url from 'node:url'
import { collectAllDependencies, getDefaultDependencyFields } from '../src/dependency'

describe('getDefaultDependencyFields', () => {
  test('basic', () => {
    expect(getDefaultDependencyFields()).toEqual([
      'dependencies',
      'optionalDependencies',
      'peerDependencies',
    ])
  })
})

describe('collectAllDependencies', () => {
  test('current repo', async () => {
    const __dirname = url.fileURLToPath(new URL('.', import.meta.url))
    const dependencies = await collectAllDependencies(path.join(__dirname, '../package.json'))

    expect(dependencies).toEqual([
      '@rollup/plugin-commonjs',
      '@rollup/plugin-json',
      '@rollup/plugin-node-resolve',
      '@rollup/plugin-typescript',
      'import-meta-resolve',
      'rollup',
      'rollup-plugin-dts',
    ])
  })

  test('nonexistent repo', async () => {
    const dependencies = await collectAllDependencies(
      locateFixtures('nonexistent/package.json'),
      undefined,
      ['rollup'],
      () => true,
    )

    expect(dependencies).toEqual([
      '@rollup/plugin-commonjs',
      '@rollup/plugin-json',
      '@rollup/plugin-node-resolve',
      '@rollup/plugin-typescript',
      'import-meta-resolve',
      'rollup',
      'rollup-plugin-dts',
    ])
  })

  test('normal repo', async () => {
    const warningDataList: unknown[] = []
    const warnSpy = jest
      .spyOn(console, 'warn')
      .mockImplementation((...args) => void warningDataList.push(...args))

    const dependencies = await collectAllDependencies(locateFixtures('normal-repo/package.json'), [
      'peerDependencies',
    ])
    expect(dependencies).toEqual(['@guanghechen/not-existed-repo', 'rollup'])

    expect(desensitize(warningDataList)).toEqual([
      "cannot find package.json for '@guanghechen/not-existed-repo'",
    ])
    warnSpy.mockRestore()
  })
})
