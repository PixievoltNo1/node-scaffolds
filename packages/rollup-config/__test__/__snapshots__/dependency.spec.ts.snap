// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectAllDependencies basic: current repo 1`] = `
[
  "@babel/code-frame",
  "@babel/helper-validator-identifier",
  "@babel/highlight",
  "@rollup/plugin-commonjs",
  "@rollup/plugin-json",
  "@rollup/plugin-node-resolve",
  "@rollup/plugin-typescript",
  "@rollup/pluginutils",
  "@types/estree",
  "@types/resolve",
  "ansi-styles",
  "balanced-match",
  "brace-expansion",
  "builtin-modules",
  "chalk",
  "color-convert",
  "color-name",
  "commondir",
  "concat-map",
  "deepmerge",
  "estree-walker",
  "fs.realpath",
  "fsevents",
  "function-bind",
  "glob",
  "has",
  "has-flag",
  "inflight",
  "inherits",
  "is-builtin-module",
  "is-core-module",
  "is-module",
  "is-reference",
  "js-tokens",
  "magic-string",
  "minimatch",
  "once",
  "path-is-absolute",
  "path-parse",
  "picomatch",
  "resolve",
  "rollup",
  "rollup-plugin-dts",
  "sourcemap-codec",
  "supports-color",
  "supports-preserve-symlinks-flag",
  "tslib",
  "typescript",
  "wrappy",
]
`;

exports[`collectAllDependencies basic: monorepo2 1`] = `
[
  "fsevents",
  "rollup",
]
`;

exports[`collectAllDependencies basic: normal-repo -- warnings 1`] = `
[
  "no such file or directory: <$WORKSPACE$>/packages/rollup-config/__test__/fixtures/normal-repo/package.json",
]
`;

exports[`collectAllDependencies basic: normal-repo 1`] = `[]`;
